From: osy <osy@turing.llc>
Date: Tue, 14 Apr 2026 14:15:49 -0700
Subject: [PATCH] virtio-gpu: Support Neptune context

Request Neptune when initializing VirGL and if neptune=true flag is set
for virtio-gpu-gl device. Neptune is a virglrenderer context type that
virtualizes D3D11/D3D12, analogous to how Venus virtualizes Vulkan.

Like Venus, Neptune requires host blob resources and the hostmem memory
window to be enabled.

Assisted-by: Claude:claude-opus-4-7

(cherry picked from utmapp/qemu commit 435e2fc057; ui/console API renamed for 11.1)
---
 docs/system/devices/virtio/virtio-gpu.rst   |  7 +++++++
 hw/display/virtio-gpu-gl.c                  |  2 ++
 hw/display/virtio-gpu-virgl.c               | 18 ++++++++++++++++++
 hw/display/virtio-gpu.c                     | 17 +++++++++++++++++
 include/hw/virtio/virtio-gpu.h              |  3 +++
 include/standard-headers/linux/virtio_gpu.h |  1 +
 meson.build                                 |  4 ++++
 tests/functional/aarch64/test_virt_gpu.py   |  2 ++
 8 files changed, 54 insertions(+)
 mode change 100755 => 100644 tests/functional/aarch64/test_virt_gpu.py

diff --git a/docs/system/devices/virtio/virtio-gpu.rst b/docs/system/devices/virtio/virtio-gpu.rst
index d05f8cb8e3f..2a269d5c031 100644
--- a/docs/system/devices/virtio/virtio-gpu.rst
+++ b/docs/system/devices/virtio/virtio-gpu.rst
@@ -177,6 +177,13 @@ of virtio-gpu host memory window. This is typically between 256M and 8G.
 
 .. _mr36814: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/36814
 
+Translation via the Neptune protocol is also supported when available in
+`virglrenderer`_. Like Venus, Neptune requires host blob support (``hostmem``
+and ``blob`` fields) and should be enabled using the ``neptune`` field.
+
+.. parsed-literal::
+    -device virtio-gpu-gl,hostmem=8G,blob=true,neptune=true
+
 virtio-gpu rutabaga
 -------------------
 
diff --git a/hw/display/virtio-gpu-gl.c b/hw/display/virtio-gpu-gl.c
index 2b7a41c4664..d333b0544ee 100644
--- a/hw/display/virtio-gpu-gl.c
+++ b/hw/display/virtio-gpu-gl.c
@@ -170,6 +170,8 @@ static const Property virtio_gpu_gl_properties[] = {
                     VIRTIO_GPU_FLAG_VENUS_ENABLED, false),
     DEFINE_PROP_BIT("drm_native_context", VirtIOGPU, parent_obj.conf.flags,
                     VIRTIO_GPU_FLAG_DRM_ENABLED, false),
+    DEFINE_PROP_BIT("neptune", VirtIOGPU, parent_obj.conf.flags,
+                    VIRTIO_GPU_FLAG_NEPTUNE_ENABLED, false),
 };
 
 static void virtio_gpu_gl_device_unrealize(DeviceState *qdev)
diff --git a/hw/display/virtio-gpu-virgl.c b/hw/display/virtio-gpu-virgl.c
index a3fc2397abf..ecc0e7483e0 100644
--- a/hw/display/virtio-gpu-virgl.c
+++ b/hw/display/virtio-gpu-virgl.c
@@ -1575,6 +1575,14 @@ static int virtio_gpu_virgl_init(VirtIOGPU *g)
         }
     }
 #endif
+#ifdef VIRGL_RENDERER_NEPTUNE
+    if (virtio_gpu_neptune_enabled(g->parent_obj.conf)) {
+        flags |= VIRGL_RENDERER_NEPTUNE;
+#ifndef CONFIG_METAL
+        flags |= VIRGL_RENDERER_RENDER_SERVER;
+#endif
+    }
+#endif
 
     ret = virgl_renderer_init(g, flags, &virtio_gpu_3d_cbs);
     if (ret != 0) {
@@ -1674,6 +1682,16 @@ GArray *virtio_gpu_virgl_get_capsets(VirtIOGPU *g)
             virtio_gpu_virgl_add_capset(capset_ids, VIRTIO_GPU_CAPSET_DRM);
         }
     }
+#ifdef VIRGL_RENDERER_NEPTUNE
+    if (virtio_gpu_neptune_enabled(g->parent_obj.conf)) {
+        virgl_renderer_get_cap_set(VIRTIO_GPU_CAPSET_NEPTUNE,
+                                   &capset_max_ver,
+                                   &capset_max_size);
+        if (capset_max_size) {
+            virtio_gpu_virgl_add_capset(capset_ids, VIRTIO_GPU_CAPSET_NEPTUNE);
+        }
+    }
+#endif
 
     return capset_ids;
 }
diff --git a/hw/display/virtio-gpu.c b/hw/display/virtio-gpu.c
index 7d4dbff3480..fe830bcb9f3 100644
--- a/hw/display/virtio-gpu.c
+++ b/hw/display/virtio-gpu.c
@@ -1603,6 +1603,10 @@ void virtio_gpu_device_realize(DeviceState *qdev, Error **errp)
     if (virtio_gpu_blob_enabled(g->parent_obj.conf)) {
 #ifdef CONFIG_METAL
         have_ext_memory = virtio_gpu_venus_enabled(g->parent_obj.conf);
+#ifdef HAVE_VIRGL_RENDERER_NEPTUNE
+        have_ext_memory = have_ext_memory ||
+            virtio_gpu_neptune_enabled(g->parent_obj.conf);
+#endif
 #else
         have_ext_memory = virtio_gpu_have_udmabuf();
 #endif
@@ -1653,6 +1657,19 @@ void virtio_gpu_device_realize(DeviceState *qdev, Error **errp)
 #endif
     }
 
+    if (virtio_gpu_neptune_enabled(g->parent_obj.conf)) {
+#ifdef HAVE_VIRGL_RENDERER_NEPTUNE
+        if (!virtio_gpu_blob_enabled(g->parent_obj.conf) ||
+            !virtio_gpu_hostmem_enabled(g->parent_obj.conf)) {
+            error_setg(errp, "neptune requires enabled blob and hostmem options");
+            return;
+        }
+#else
+        error_setg(errp, "virglrenderer does not support neptune");
+        return;
+#endif
+    }
+
     if (!virtio_gpu_base_device_realize(qdev,
                                         virtio_gpu_handle_ctrl_cb,
                                         virtio_gpu_handle_cursor_cb,
diff --git a/include/hw/virtio/virtio-gpu.h b/include/hw/virtio/virtio-gpu.h
index 220231ec9d4..1f552ef69de 100644
--- a/include/hw/virtio/virtio-gpu.h
+++ b/include/hw/virtio/virtio-gpu.h
@@ -101,6 +101,7 @@ enum virtio_gpu_base_conf_flags {
     VIRTIO_GPU_FLAG_VENUS_ENABLED,
     VIRTIO_GPU_FLAG_RESOURCE_UUID_ENABLED,
     VIRTIO_GPU_FLAG_DRM_ENABLED,
+    VIRTIO_GPU_FLAG_NEPTUNE_ENABLED,
 };
 
 #define virtio_gpu_virgl_enabled(_cfg) \
@@ -125,6 +126,8 @@ enum virtio_gpu_base_conf_flags {
     (_cfg.flags & (1 << VIRTIO_GPU_FLAG_VENUS_ENABLED))
 #define virtio_gpu_drm_enabled(_cfg) \
     (_cfg.flags & (1 << VIRTIO_GPU_FLAG_DRM_ENABLED))
+#define virtio_gpu_neptune_enabled(_cfg) \
+    (_cfg.flags & (1 << VIRTIO_GPU_FLAG_NEPTUNE_ENABLED))
 
 struct virtio_gpu_base_conf {
     uint32_t max_outputs;
diff --git a/include/standard-headers/linux/virtio_gpu.h b/include/standard-headers/linux/virtio_gpu.h
index c67f50f56cd..474d532a413 100644
--- a/include/standard-headers/linux/virtio_gpu.h
+++ b/include/standard-headers/linux/virtio_gpu.h
@@ -321,6 +321,7 @@ struct virtio_gpu_cmd_submit {
 #define VIRTIO_GPU_CAPSET_VENUS 4
 #define VIRTIO_GPU_CAPSET_CROSS_DOMAIN 5
 #define VIRTIO_GPU_CAPSET_DRM 6
+#define VIRTIO_GPU_CAPSET_NEPTUNE 7
 
 /* VIRTIO_GPU_CMD_GET_CAPSET_INFO */
 struct virtio_gpu_get_capset_info {
diff --git a/meson.build b/meson.build
index c2edc3aaf41..120d56f1ac2 100644
--- a/meson.build
+++ b/meson.build
@@ -2523,6 +2523,10 @@ if virgl.found()
   config_host_data.set('VIRGL_VERSION_MAJOR', virgl.version().split('.')[0])
   config_host_data.set('VIRGL_VERSION_MINOR', virgl.version().split('.')[1])
   config_host_data.set('VIRGL_VERSION_MICRO', virgl.version().split('.')[2])
+  config_host_data.set('HAVE_VIRGL_RENDERER_NEPTUNE',
+                        cc.has_header_symbol('virglrenderer.h',
+                                            'VIRGL_RENDERER_NEPTUNE',
+                                            dependencies: virgl))
 endif
 config_host_data.set('CONFIG_VIRTFS', have_virtfs)
 config_host_data.set('CONFIG_VTE', vte.found())
diff --git a/tests/functional/aarch64/test_virt_gpu.py b/tests/functional/aarch64/test_virt_gpu.py
old mode 100755
new mode 100644
index 6ed0e0fc1a9..5ae7ba962e4
--- a/tests/functional/aarch64/test_virt_gpu.py
+++ b/tests/functional/aarch64/test_virt_gpu.py
@@ -66,6 +66,8 @@ def _launch_virt_gpu(self, gpu_device):
                 self.skipTest("No blob support for virtio-gpu")
             elif "old virglrenderer, venus unsupported" in excp.output:
                 self.skipTest("No venus support for virtio-gpu")
+            elif "virglrenderer does not support neptune" in excp.output:
+                self.skipTest("No neptune support for virtio-gpu")
             elif "egl: no drm render node available" in excp.output:
                 self.skipTest("Can't access host DRM render node")
             elif "'type' does not accept value 'egl-headless'" in excp.output:
